support/e2e-test

ryu-sato opened the pull request:
Reivew & merge task
* https://weseek.myjetbrains.com/youtrack/issue/INFRA-274

docker-compose を使った e2e テストを追加

ryu-sato committed:
docker-compose を使った e2e テストを追加

skomma commented on Dockerfile.test:
```diff
@@ -0,0 +1,31 @@
+FROM alpine:3.7
+
+LABEL maintainer="WESEEK <info@weseek.co.jp>"
```
maintainer は自信を持って自分の名前で！
あと、description がほしい。

ryu-sato replied:
本体の Dockerfile コピペしただけなのですが。 :-P
適宜修正しておきます。

skomma commented on Dockerfile.test:
```diff
@@ -0,0 +1,31 @@
+    py2-pip \
+    mongodb-tools \
+    curl
+RUN pip install awscli
```
e2e にするなら、この辺は base の Dockerfile から作った Docker image のものを使いたいなぁ。

ryu-sato replied:
そう。そうなんですが aws コマンドを s3proxy に向けられず、苦慮した結果こうしました。
本体　base の Dockerfile を修正して、本番かテストかで分岐させるとかですかね。

skomma commented on test/Dockerfile.util:
```diff
@@ -0,0 +1,17 @@
+FROM alpine:3.7
+
+LABEL maintainer="WESEEK <info@weseek.co.jp>"
```
ここも。

skomma commented on test/docker-compose.yml:
```diff
@@ -0,0 +1,94 @@
+    build:
+      context: ../
+      dockerfile: Dockerfile.test
+    depends_on:
```
depends_on とか links はもはや不要なはず。
どうせ、dockerize で待つし。

http://docs.docker.jp/compose/startup-order.html

ryu-sato replied:
dockerize で待つので depends_on が要らないことは納得しましたが、links は mongodb に接続するために必要と思ったので要否確認してから適宜修正します。

ryu-sato replied:
docker-compose yml version 2 から links を設定しなくても service 名でアクセスできるようになったのですね。
今回は version 3 なので不要と理解しました。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash -e
+
+COMPOSE_PROJECT_NAME=mongodbawesomebackuptest
```
これは .env に移動したいところ。
https://docs.docker.com/compose/env-file/

ryu-sato replied:
test 配下の .env に追加してうまく行かなかったのでそうしました。(テスト内で変数としても使っているので)
.env に移動できるかもう一度検討してみます。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,71 @@
+TODAY=${TODAY} \
+  docker-compose -p ${COMPOSE_PROJECT_NAME} -f test/docker-compose.yml up -d --build
+
+# sleep while backup is executed in cron mode (at leat 60s)
```
least

skomma replied:
なぜ 60 秒待つのかコメントが欲しい。

ryu-sato replied:
追記します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,71 @@
+check_s3_file_exist ${S3_ENDPOINT_URL} "app_backup_cronmode/backup-${TODAY}.tar.bz2"
+
+TESTING_CONTAINER=""
+echo "***** ALL TEST ARE SUCCESS *****"
```
ALL TESTS ARE SUCCESSFUL だな

ryu-sato replied:
修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,71 @@
+TESTING_CONTAINER="app_restore"
+## should upload file `backup-#{TODAY}.tar.bz2` to S3
+check_s3_file_exist ${S3_ENDPOINT_URL} "app_restore/backup-${TODAY}.tar.bz2"
+## should exit code is 0
```
exit code should be 0

ryu-sato replied:
修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,71 @@
+cd ..
+
+TODAY=${TODAY} \
+  docker-compose -p ${COMPOSE_PROJECT_NAME} -f test/docker-compose.yml up -d --build
```
-d せずに `&` でバックグラウンドで動かしたらログが出てくるしいいんじゃないかしら？

ryu-sato replied:
なるほど。試して適宜修正しますね。

ryu-sato committed:
correspond F

ryu-sato committed:
correspond FB

ryu-sato committed:
add permission to exec

ryu-sato committed:
correspond FB. modify dump_all_log to get container list correctory

ryu-sato committed:
change working directory of e2e test

ryu-sato committed:
correspond FB

ryu-sato committed:
remove unused dump_all_log function, and set environment AWSCLIOPT

ryu-sato committed:
set .env variables

ryu-sato committed:
Merge branch 'master' into support/e2e-test

ryu-sato commented on .gitignore:
```diff
@@ -1,2 +1,2 @@
 # for Docker
-.env
+./.env
```
test 配下の .env ファイルが repository 管理できていなかったため修正。

skomma replied:
頭のピリオドいらなくない？

ryu-sato replied:
なるほど。確認して適宜修正します。

ryu-sato commented on Dockerfile:
```diff
@@ -7,9 +7,15 @@ RUN apk add --no-cache \
+    curl
 RUN pip install awscli
 
+# install dockerize(it is needed by e2e test)
```
e2e test で対象とする Dockerfile を本番で利用するものと同じにするため dockerize をインストールすることにしました。

ryu-sato commented on bin/functions.sh:
```diff
@@ -3,6 +3,7 @@ AWSCLI="/usr/bin/aws"
 AWSCLI_COPY_OPT="s3 cp"
 AWSCLI_LIST_OPT="s3 ls"
 AWSCLI_DEL_OPT="s3 rm"
+AWSCLIOPT=${AWSCLIOPT:-}
```
e2e test 時に awscli コマンドの接続先を変更したいため AWSCLIOPT を追加しました。
※ 初期値は空なので記載する必要はありませんが、変数を内部で利用していることが分かりやすいように記載しています。

skomma commented on test/Dockerfile.util:
```diff
@@ -0,0 +1,18 @@
+FROM alpine:3.7
```
この Dockerfile いる？
ルートの Dockerfile で事足りるよね？

ryu-sato replied:
Dockerfile.util (仮名) の役割が変わりそうなのでこのコメントは一旦置いておきます。

skomma commented on test/docker-compose.yml:
```diff
@@ -0,0 +1,82 @@
+       -timeout 60s /bin/bash -c
+        'curl --request PUT http://s3proxy:80/app_default 
+         && curl --request PUT http://s3proxy:80/app_backup_cronmode
+         && curl --request PUT http://s3proxy:80/app_restore'"
```
このコンテナ、まとめないで各コンテナで起動時にやればいいんじゃない？

ryu-sato replied:
各コンテナで CMD に設定したとしても entrypoint.sh の引数になるだけで ${CMD}.sh を呼び出そうとして失敗すると思うのですが、コンテナ起動時に実行できるのでしょうか？

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,64 @@
+trap handle_exit EXIT
+trap 'rc=$?; trap - EXIT; handle_exit; exit $?' INT PIPE TERM
+
+# check a S3 file is exist
```
動詞が2つある

ryu-sato replied:
修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,64 @@
+# sleep because test backup is executed every minutes in cron mode
+sleep 65
+
+# app_default
```
コメントがここだけ違う…

ryu-sato replied:
気が付きませんでした。修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,64 @@
+# Test for app_restore
+TESTING_CONTAINER="app_restore"
+## should upload file `backup-#{TODAY}.tar.bz2` to S3
+check_s3_file_exist ${S3_ENDPOINT_URL} "app_restore/backup-${TODAY}.tar.bz2"
```
restore の確認になってない？

ryu-sato replied:
確かに restore 出来たかどうかは確認できていないです。よい方法が浮かばなかったもので。
ちょっと考えます。

ryu-sato committed:
Merge branch 'master' into support/e2e-test

ryu-sato committed:
FB対応

ryu-sato committed:
COMPOSE_PROJECT_NAMEを戻戻し

ryu-sato committed:
mongodb のコンテナ名を修正

ryu-sato committed:
init_s3proxy_1の終了をdocker waitで待つよう修正

ryu-sato committed:
docker wait を実行するとシェルが終了してしまったためwhileで待つよう修正

ryu-sato committed:
s3proxy を docker から link して接続。

ryu-sato committed:
while の終了条件を修正

ryu-sato committed:
コメント追記

ryu-sato committed:
docker のコンテナIDを正しく取得できるよう修正

ryu-sato committed:
docker rm は不要なので削除

ryu-sato committed:
docker image のタグを指定する変数の名前を修正
テストプログラムに対して設定できる変数をファイル上部に追記

ryu-sato committed:
docker run の共通オプションを変数にして再利用

ryu-sato committed:
envファイルの読み込み位置を修正

ryu-sato committed:
CONTAINER_ID のテストを行う箇所のカッコの位置を修正

ryu-sato committed:
FB対応。e2e.sh に書かれていた docker run の内容を docker-compose.yml に集約するよう修正。

ryu-sato committed:
テスト用 docker-compose を実行する記述の誤りを修正

ryu-sato committed:
external_link と depends_on を削除 (下記エラーが出て test 用の docker-compose は起動しない)
***** TEST FAILED *****
WARNING: Found orphan containers (mongodbawesomebackuptest_mongo_1, mongodbawesomebackuptest_s3proxy_1) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up.

ryu-sato committed:
分割していた e2e テスト用 docker-compose.yml ファイルを1つに集約

ryu-sato committed:
build オプションの位置を修正

ryu-sato committed:
mongodb が起動するまで dockerize で待機するよう修正。合わせてコンテナ名を修正。

ryu-sato committed:
コンテナ名を修正

ryu-sato committed:
docker-compose 実行処理のオプションを整理

ryu-sato committed:
TODAY 変数を注入

ryu-sato committed:
docker-compose up が終了すると handle_exit している様子があるのでテスト実行時に docker-compose down するよう修正

ryu-sato committed:
バックグラウンドで docker-compose up を実行する必要のない箇所はフォアグラウンドで実行するよう修正し、終了を待機する処理は廃止。

ryu-sato committed:
バックグラウンドで処理する必要があったので元戻し

ryu-sato committed:
実行するテスト内容を echo するよう修正

ryu-sato committed:
docker container を待機せずに check_s3_file_exist を実行して失敗していたため待機する関数を追加しつつ、待機する処理を追加

ryu-sato committed:
wait_docker_container に渡すパラメータを文字列として指定

ryu-sato committed:
stop_docker_container に渡すパラメータを修正

ryu-sato committed:
テストが完了したら docker-compose down を実行するよう修正

ryu-sato committed:
テスト成功時のメッセージを出力する箇所を処理の最終行へ移動

ryu-sato committed:
テスト実行時のメッセージを整形

ryu-sato committed:
メッセージ出力内容を修正

ryu-sato committed:
テスト結果を出力するメッセージを修正

ryu-sato committed:
コメントの英語を修正

ryu-sato committed:
add dummy backup file

ryu-sato committed:
Merge branch 'support/e2e-test' of https://github.com/weseek/mongodb-awesome-backup into support/e2e-test

ryu-sato committed:
replace dummy backup file for test

ryu-sato committed:
restore を行う際に利用する dummy ファイルの名前を修正
bucket 作成と cp を行う際に aws コマンドを使うよう修正

ryu-sato committed:
Dockerfile.util のファイル名を実態に合わせて修正

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+TODAY=`/bin/date +%Y%m%d` # It is used to generate file name to restore
+
+# Clean up before start s3proxy and mongodb
+docker-compose down
```
-v 追加で。
mongodb とかデフォルトで volume を残すけど、docker-compose.yml に明記しなくても消えるのか要チェック。

ryu-sato replied:
確認して修正します。

ryu-sato replied:
volume は残ってました。 `-v` オプションをつけることにします。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+echo 'Finished test for app_backup_cronmode: OK'
+
+# Clean up all containers
+docker-compose down
```
-v 追加で。

ryu-sato replied:
確認して修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+  done
+}
+
+# Wait while container exist
```
コメント要修正

ryu-sato replied:
適宜修正します。

skomma commented on test/docker-compose.yml:
```diff
@@ -0,0 +1,63 @@
+      - TZ=Asia/Tokyo
+
+  # initialize S3 bucket for test
+  init_s3proxy_and_mongo:
```
init でいいんじゃない？

ryu-sato replied:
出来る限り説明した方がよいと思いましたが、伝わるようなら短い方がよいので修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+#   ARGS
+#     $1 ... ENDPOINT_URL: Endpoint URL of S3
+#     $2 ... S3_FILE_PATH: File path of S3 to be checked for existence
+check_s3_file_exist() {
```
check... はよくない。
https://qiita.com/KeithYokoma/items/2193cf79ba76563e3db6

存在していることを強制したいのであれば、`assert_file_exists_on_s3` とか？

そういえば、この関数は return してるだけだけど、テスト失敗した場合は途中で終了とかしないでいいの？

ryu-sato replied:
確かに何を check するのか分からないですね。(期待する動作が分からない)
修正します。

> そういえば、この関数は return してるだけだけど、テスト失敗した場合は途中で終了とかしないでいいの？

function の return が 0 以外であれば bash -e で終了するのでその点は大丈夫です。

skomma commented on test/.env:
```diff
@@ -0,0 +1,4 @@
+COMPOSE_PROJECT_NAME=mongodbawesomebackuptest
```
これ、docker-compose が読み込む `.env` として必要なのは、↑の行だけだよね？
コンテナ内でだけ必要な、↓3行は別ファイルに分離してほしいかなぁ。

ryu-sato replied:
いや、docker-compose が読み込む .env で4行すべて必要です。
e2e.sh では COMPOSE_PROJECT_NAME だけ流用してますが、そっちのことを言ってますか？

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+
+  ENDPOINT_URL=$1
+  S3_FILE_PATH=$2
+  curl -I -L --silent "${ENDPOINT_URL}/${S3_FILE_PATH}" 2>&1 | grep -e '^HTTP/' | grep -q '200 OK'
```
末尾の grep は `'^HTTP/.\+200 OK'` にまとめられるね。

ryu-sato replied:
そうですね。まとめます。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+# Expect for app_backup_cronmode
+## stop container
+##   before stop, sleep 65s because test backup is executed every minute in cron mode
+stop_docker_container "app_backup_cronmode" "-t 65"
```
docker-compose stop じゃダメなのかい？

```
$ docker-compose stop --help
Stop running containers without removing them.

They can be started again with `docker-compose start`.

Usage: stop [options] [SERVICE...]

Options:
  -t, --timeout TIMEOUT      Specify a shutdown timeout in seconds.
                             (default: 10)
```

ryu-sato replied:
ああ、行けそうですね。修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+wait_docker_container "app_restore"
+## should restored mongodb
+DUMMY_RECORD_NUM=$(docker-compose exec mongo bash -c 'echo -e "use dummy;\n db.dummy.find({name: \"test\"})\n" | mongo | grep "ObjectId" | wc -l')
+if [ "x${DUMMY_RECORD_NUM}" = "x1" ]; then echo 'FAILED'; exit 1; fi
```
↑2行は関数化したほうがよいのでは？

ryu-sato replied:
assert～ といった関数名で作成しておきます。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,101 @@
+docker-compose up --build init_s3proxy_and_mongo s3proxy mongo &
+
+# Sleep while s3 bucket is created
+wait_docker_container "init_s3proxy_and_mongo"
```
`wait_docker_container` のところってコンテナが終了するまで待つんだよね？

`-d` なしの `docker-compose up init` で終了したら戻ってこないかしら？

ryu-sato replied:
戻ってくる気がしますね。確認して適宜修正します。

ryu-sato replied:
戻ってきませんでした。

ryu-sato replied:
考えてみれば init が終了しても s3proxy や mongo は起動したままなのでフォアグラウンドで実行することはできませんでしたね。

ryu-sato committed:
docker-compose down 時に volume を削除するようオプションを追加

ryu-sato committed:
docker コマンドは使わず docker-compose stop を使うよう修正

ryu-sato committed:
container 名を修正

ryu-sato committed:
container 名を修正

ryu-sato committed:
check という単語で関数を作らないよう修正

ryu-sato committed:
env ファイルを docker-compose が必要なものか container の動作に必要なものかで分離。

ryu-sato committed:
grep を集約

ryu-sato committed:
restore 成功可否の確認を関数化

ryu-sato committed:
assert_file_exists_on_s3 が失敗した時にエラーメッセージを表示

ryu-sato committed:
不要な wait_docker_container を実行しないように修正

ryu-sato committed:
.env を使わないようにした変更に伴う修正

ryu-sato committed:
wait_docker_container が必要だったため元戻し

ryu-sato committed:
使っていない変数を削除

ryu-sato committed:
environment が yaml により import されなかったので修正

ryu-sato committed:
restore 成否確認の判定式が誤っていたため修正

ryu-sato committed:
docker container 名を確認するロジックを前方一致しづらいように修正

ryu-sato committed:
テスト失敗時のメッセージを追加

ryu-sato committed:
assert_file_exists_on_s3 の変数名を修正。
使っていない変数を削除。

ryu-sato committed:
assert_restore_dummy_record 関数を修正

ryu-sato committed:
関数名を修正

ryu-sato committed:
終了時に docker-compose down を実行するハンドラを実装

skomma commented on test/docker-compose.yml:
```diff
@@ -0,0 +1,71 @@
+  app_backup_cronmode:
+    <<: *app_default
+    environment:
+      - AWS_ACCESS_KEY_ID=testid
```
yaml の参照を使った List のマージってできないんだっけ？
マージできないから全部書き直してる系？

ryu-sato replied:
app_default のマージでは反映されませんでした。
environment に名前を付けてマージできるかどうかは試していないので試してみます。

ryu-sato replied:
environment にアンカー付けてマージできたので修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,95 @@
+  docker-compose down -v
+}
+trap handle_exit EXIT
+trap 'rc=$?; trap - EXIT; handle_exit; exit $?' INT PIPE TERM
```
参考にしたページがあるならコメントで書いておいて。

ryu-sato replied:
追加します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,95 @@
+  if [ "x${HTTP_OK}" = "x" ]; then echo 'assert_file_exists_on_s3 FAILED'; exit 1; fi
+}
+
+# assert restore successful
```
restore **is** successful

ryu-sato replied:
修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,95 @@
+docker-compose up --build init s3proxy mongo &
+
+# Wait while init exist
+wait_docker_container "init"
```
`docker-compose run init` でも init 単体で待てなかった？

ryu-sato replied:
やってみます。

ryu-sato replied:
run でも up でも事前に s3proxy, mongo を立ち上げた後に init 単体で指定して FG で実行すれば、実行完了を待つことはできましたが、下記エラーとなりました。
sleep 3 を挟むと成功するため、恐らく network が初期化される前に docker-compose up を連続実行しているのが問題だと思われます。

```
  : <snip>
mongodbawesomebackuptest_init_1 exited with code 0
ERROR: 2 matches found based on name: network mongodbawesomebackuptest_default is ambiguous
  : <snip>
```

ryu-sato committed:
参考にしたサイトをコメントで追記

ryu-sato committed:
環境変数を DRY に記述

ryu-sato committed:
refs: INFRA-273 英語修正

ryu-sato committed:
refs: INFRA-273 FB対応

ryu-sato committed:
sleep を適宜追加

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,76 @@
+
+# Start s3proxy and mongodb
+docker-compose up --build s3proxy mongo &
+sleep 3 # wait for that network of docker-compose is ready
```
`wait for the network of docker-compose to be ready`

ryu-sato replied:
修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,76 @@
+echo 'Finished test for app_default: OK'
+
+# Expect for app_restore
+docker-compose up --build app_restore
```
ここ以降の `docker-compose up` には `--build` はいらないんじゃないか？
app_default と同じイメージを使うのよね？

ryu-sato replied:
同じイメージを使うので要らないですね。消しておきます。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,76 @@
+
+# Expect for app_backup_cronmode
+docker-compose up --build app_backup_cronmode &
+sleep 3 # wait for that network of docker-compose is ready
```
こっちのコメントも要修正

ryu-sato replied:
修正します。

skomma commented on test/e2e.sh:
```diff
@@ -0,0 +1,76 @@
+sleep 3 # wait for that network of docker-compose is ready
+## stop container
+##   before stop, sleep 65s because test backup is executed every minute in cron mode
+docker-compose stop -t 65 app_backup_cronmode
```
ん～ここの `-t` は↑の sleep と一緒にしちゃっていいんじゃない？

ryu-sato replied:
> ん～ここの -t は↑の sleep と一緒にしちゃっていいんじゃない？

これ、ダメだったんです。。。
docker-compose up -> stop を sleep 入れずに実行するとエラー (network がほげほげ)
なので、ここにも sleep は入れる必要ありそうです。

ryu-sato replied:
-t 65 は削除して、sleep の方にまとめて sleep 65 にする方針で行きます。
待機する理由はコメントに記載します。

ryu-sato committed:
refs: INFRA-274 FB対応
